// ng-dynamic-autocomplete-searchfield/bundles/dynamicAutocompleteSearchField.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/platform-browser'), require('@angular/common'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/platform-browser', '@angular/common', '@angular/forms'], factory) :
    (factory((global.ng = global.ng || {}, global.ng.dynamicAutocompleteSearchField = {}),global.ng.core,global.platformBrowser,global.common,global.forms));
}(this, (function (exports,core,platformBrowser,common,forms) { 'use strict';

    var SearchboxComponent = /** @class */ (function () {
        function SearchboxComponent() {
            this.changeEvent = new core.EventEmitter();
            this.placeholder = 'search';
            this.noRecord = 'no data found';
            this.status = true;
        }
        SearchboxComponent.prototype.ngOnInit = function () {
            if (this.options.searchData.ishighlight) {
                this.highlightColour = 'yellow';
                if (this.options.searchData.highlightColour) {
                    this.highlightColour = this.options.searchData.highlightColour;
                }
            }
            var defautSearchBy = [];
            for (var _i = 0, _a = this.options.data; _i < _a.length; _i++) {
                var obj = _a[_i];
                Object.keys(obj).forEach(function (data) {
                    if (defautSearchBy.indexOf(data) === -1) {
                        defautSearchBy.push(data);
                    }
                });
            }
            this.searchBy = defautSearchBy;
            this.displayBy = defautSearchBy;
            if (this.options.searchData.placeHolder) {
                this.placeholder = this.options.searchData.placeHolder;
            }
            if (this.options.searchData.noRecord) {
                this.noRecord = this.options.searchData.noRecord;
            }
            if (this.options.searchData.searchBy) {
                if (this.options.searchData.searchBy.length > 0) {
                    this.searchBy = this.options.searchData.searchBy;
                }
            }
            if (this.options.searchData.displayBy) {
                if (this.options.searchData.displayBy.length > 0) {
                    this.displayBy = this.options.searchData.displayBy;
                }
            }
        };
        SearchboxComponent.prototype.searchRowValue = function (value) {
            var selectedValue = [];
            this.options.searchData.displayBy.forEach(function (key) {
                selectedValue.push(value[key]);
            });
            this.status = !this.status;
            this.searchText = selectedValue.join(' ');
            this.changeEvent.emit(value);
        };
        SearchboxComponent.prototype.checkNoRecord = function (data) {
            if (data.length === 0) {
                if (this.status) {
                    return true;
                }
            }
            else {
                return false;
            }
        };
        SearchboxComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'app-searchbox',
                        templateUrl: './searchbox.component.html',
                        styleUrls: ['./searchbox.component.css']
                    },] },
        ];
        /** @nocollapse */
        SearchboxComponent.ctorParameters = function () { return []; };
        SearchboxComponent.propDecorators = {
            options: [{ type: core.Input }],
            changeEvent: [{ type: core.Output }]
        };
        return SearchboxComponent;
    }());

    var SearchPipe = /** @class */ (function () {
        function SearchPipe() {
        }
        SearchPipe.prototype.transform = function (value, keys, term) {
            if (!term) {
                return value;
            }
            return (value || []).filter(function (item) { return keys.some(function (key) { return item.hasOwnProperty(key) && new RegExp(term, 'gi').test(item[key]); }); });
        };
        SearchPipe.decorators = [
            { type: core.Pipe, args: [{
                        name: 'searchFilter'
                    },] },
        ];
        return SearchPipe;
    }());

    var HighlightSearch = /** @class */ (function () {
        function HighlightSearch(_sanitizer) {
            this._sanitizer = _sanitizer;
        }
        HighlightSearch.prototype.transform = function (value, args, getColour) {
            var re = new RegExp(args, 'gi');
            var string = value.toString().replace(re, "<span style = \"background-color:" + getColour + "\">" + args + "</span>");
            return this._sanitizer.bypassSecurityTrustHtml(string);
        };
        HighlightSearch.decorators = [
            { type: core.Pipe, args: [{
                        name: 'highlight',
                    },] },
        ];
        /** @nocollapse */
        HighlightSearch.ctorParameters = function () { return [
            { type: platformBrowser.DomSanitizer }
        ]; };
        return HighlightSearch;
    }());

    var AutocompleteSearchModule = /** @class */ (function () {
        function AutocompleteSearchModule() {
        }
        AutocompleteSearchModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule,
                            forms.FormsModule
                        ],
                        declarations: [SearchboxComponent,
                            SearchPipe,
                            HighlightSearch
                        ],
                        exports: [
                            SearchboxComponent, SearchPipe, HighlightSearch
                        ]
                    },] },
        ];
        return AutocompleteSearchModule;
    }());

    exports.AutocompleteSearchModule = AutocompleteSearchModule;
    exports.SearchboxComponent = SearchboxComponent;
    exports.SearchPipe = SearchPipe;
    exports.HighlightSearch = HighlightSearch;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
