// angular-multi-single-dropdown/bundles/amazing.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/forms'), require('@angular/platform-browser')) :
    typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', '@angular/forms', '@angular/platform-browser'], factory) :
    (factory((global.ng = global.ng || {}, global.ng.amazing = {}),global.ng.core,global.common,global.forms,global.platformBrowser));
}(this, (function (exports,core,common,forms,platformBrowser) { 'use strict';

    var SearchPipe = /** @class */ (function () {
        function SearchPipe() {
        }
        SearchPipe.prototype.transform = function (items, searchItem, option) {
            if (searchItem) {
                return items.filter(function (x) {
                    return x[option.textKey].indexOf(searchItem) > -1;
                });
            }
            else {
                return items;
            }
        };
        SearchPipe.decorators = [
            { type: core.Pipe, args: [{
                        name: 'search'
                    },] },
        ];
        return SearchPipe;
    }());

    var DropdownComponent = /** @class */ (function () {
        function DropdownComponent() {
            this.data = [];
            this.toggle = true;
            this.messageEvent = new core.EventEmitter();
        }
        DropdownComponent.prototype.ngOnInit = function () {
            var _this = this;
            this.option.array.forEach(function (v) {
                if (v.checked) {
                    _this.data.push(v);
                }
            });
        };
        DropdownComponent.prototype.getData = function (event, item) {
            if (event.target.checked) {
                this.data.push(item);
            }
            else {
                if (this.data) {
                    this.index = this.data.findIndex(function (x) {
                        return Object.keys(item).every(function (key) {
                            return x[key] === item[key];
                        });
                    });
                    this.data.splice(this.index, 1);
                }
            }
            this.messageEvent.emit(this.data);
        };
        DropdownComponent.prototype.selectAll = function (getArr) {
            var _this = this;
            getArr.forEach(function (element) {
                if (element.checked === true) {
                    return false;
                }
                else {
                    element.checked = true;
                    _this.data.push(element);
                }
            });
            this.messageEvent.emit(this.data);
        };
        DropdownComponent.prototype.unSelectAll = function (arrData) {
            var _this = this;
            arrData.forEach(function (element) {
                if (element.checked === false) {
                    return true;
                }
                else {
                    element.checked = false;
                    _this.index = _this.data.findIndex(function (x) {
                        return Object.keys(element).every(function (key) {
                            return x[key] === element[key];
                        });
                    });
                    _this.data.splice(_this.index, 1);
                }
            });
            this.messageEvent.emit(this.data);
        };
        DropdownComponent.prototype.updateSelection = function (position, array, event, item) {
            this.data = [];
            array.forEach(function (x, index) {
                if (position !== index) {
                    x.checked = false;
                }
            });
            if (event.target.checked) {
                this.data.push(item);
            }
            this.messageEvent.emit(this.data);
        };
        DropdownComponent.prototype.clear = function () {
            this.searchText = '';
        };
        DropdownComponent.prototype.delete = function (item) {
            this.index = this.data.findIndex(function (x) {
                return Object.keys(item).every(function (key) {
                    return x[key] === item[key];
                });
            });
            this.data.splice(this.index, 1);
            var check = this.option.array.forEach(function (element) {
                Object.keys(item).every(function (key) {
                    if (element[key] === item[key]) {
                        return element.checked = false;
                    }
                });
            });
            this.toggle = true;
        };
        DropdownComponent.prototype.closeDropdown = function (flag) {
            this.toggle = true;
        };
        DropdownComponent.prototype.manageDrop = function (flag) {
            this.toggle = flag;
        };
        DropdownComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'app-dropdown',
                        templateUrl: './dropdown.component.html',
                        styleUrls: ['./dropdown.component.css']
                    },] },
        ];
        /** @nocollapse */
        DropdownComponent.ctorParameters = function () { return []; };
        DropdownComponent.propDecorators = {
            option: [{ type: core.Input }],
            messageEvent: [{ type: core.Output }]
        };
        return DropdownComponent;
    }());

    var OutsideDirective = /** @class */ (function () {
        function OutsideDirective(elementRef) {
            this.elementRef = elementRef;
            this.flag = false;
            this.closeSection = new core.EventEmitter();
            this.appOutside = new core.EventEmitter();
        }
        OutsideDirective.prototype.onDocumentClick = function (event) {
            this.targetElement = event.target;
            this.flag = false;
            if (this.targetElement && !this.elementRef.nativeElement.contains(this.targetElement)) {
                if ((this.flag === false) && (this.targetElement.className === 'buttons')) {
                    this.flag = true;
                    this.closeSection.emit(this.flag);
                    return false;
                }
                else if ((this.flag) && (this.targetElement.className === 'buttons')) {
                    this.flag = false;
                    this.closeSection.emit(this.flag);
                    return false;
                }
                else {
                    this.flag = true;
                    this.appOutside.emit(this.flag);
                }
            }
        };
        OutsideDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[appOutside]'
                    },] },
        ];
        /** @nocollapse */
        OutsideDirective.ctorParameters = function () { return [
            { type: core.ElementRef }
        ]; };
        OutsideDirective.propDecorators = {
            closeSection: [{ type: core.Output }],
            appOutside: [{ type: core.Output }],
            onDocumentClick: [{ type: core.HostListener, args: ['document:click', ['$event'],] }]
        };
        return OutsideDirective;
    }());

    var multiSingleDropDownModule = /** @class */ (function () {
        function multiSingleDropDownModule() {
        }
        multiSingleDropDownModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule,
                            forms.FormsModule,
                            platformBrowser.BrowserModule
                        ],
                        declarations: [DropdownComponent, SearchPipe, OutsideDirective],
                        exports: [
                            DropdownComponent, SearchPipe, OutsideDirective
                        ]
                    },] },
        ];
        return multiSingleDropDownModule;
    }());

    exports.multiSingleDropDownModule = multiSingleDropDownModule;
    exports.SearchPipe = SearchPipe;
    exports.DropdownComponent = DropdownComponent;
    exports.OutsideDirective = OutsideDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
